Added Non-Functional Changes

yogendra opened the pull request:
Added testing support and test case for generate.js
Added editor config, eslint and editorconfig

yogendra committed:
Added Testing, Linting, npmignore & Editorconfig

yogendra committed:
Added support for VS Code for Debugging

vot commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
 node_modules
 npm-debug.log
-
+coverage
```
Could you potentially move the relevant contents of `.npmignore` to `.gitignore`? npm by default inherits contents of `.gitignore` and it's easier to maintain a single list where possible.

I think `.vscode`, `coverage` and `yarn.lock` could be placed in `.gitignore`. `.npmignore` could go then.
I see no harm in having tests and lint configuration in the package but definitely we need readme. I'd also rather not add editor-specific configurations to the repo.

yogendra replied:
I like to maintain single ignore file too. 
But, as per [this](https://yarnpkg.com/en/docs/yarn-lock#toc-check-into-source-control) `yarn.lock` should be checked in. This ensures that anyone running yarn will get exact set of transitive dependencies (like npm shrinkwrap). 
`.vscode` is ide configuration for debugging, so I thought it would make it handy to quickly debug test/source. 
but neither of these needs to be published on NPM, so you will still have to maintain `.npmignore`

yogendra replied:
Also, I thought we can reduce module size by only including required files and no tests. That's why I split gitignore and npmignore

vot replied:
I don't think there's need for that level of shrink-wrapping - the only module currently required is `clarg` which is another thing I maintain and it has no dependencies of its own either (that's the entire point of it actually - it's a simple CLI argument parser).

While it could potentially affect tests I don't think that should be enough reason to add it to the repo quite honestly. I'm really inclined to just carry on with the simpler approach of single ignore file (as for tests I think there's no harm in including them in the actual package).

yogendra replied:
Ok sold!

vot commented on package.json:
```diff
@@ -17,10 +17,7 @@
-    "type": "git",
-    "url": "https://github.com/vot/xkpasswd-node.git"
-  },
+  "repository": "https://github.com/vot/xkpasswd-node.git",
```
Never paid much attention to the format of this in other `package.json` files. Is this a valid shorthand?

I'm looking at https://docs.npmjs.com/getting-started/using-a-package.json where they simply use the old format so I'd rather stick to that but I'm curious, is the one-line repository a thing now?

yogendra replied:
i think this came up in some linter (may be), hence I split it. But I am okay both ways.

vot replied:
Hmm cool. I'd rather just go with the documented npm way, just for the sake of ensuring it doesn't cause any weird `package.json` issues in older versions or private/cache servers.

yogendra replied:
Agree.

vot commented on .eslintrc.json:
```diff
@@ -0,0 +1,20 @@
+        "browser": true,
+        "commonjs": true,
+        "es6": true,
+        "node": true
```
I haven't checked out your branch yet to test it myself but are these settings correct? I think only `node` should be set to true here, no ES6 features are used that I can think of and it wouldn't run in a browser as-is.

yogendra replied:
Agree.

vot commented:
Hey buddy,
This is really amazing, thanks for getting this module tested and linted.
I've added a few comments in the annotations, let me know if they sound reasonable.

Thank you for putting time into this :)

yogendra commented:
@vot , I can't see the annotations in the files. Can you tell me which commit/file is that?

vot commented:
@yogendra - I just realised I had to submit the review in order for you to see it. You should now see the comments.

yogendra commented on package.json:
```diff
@@ -17,10 +17,7 @@
-    "type": "git",
-    "url": "https://github.com/vot/xkpasswd-node.git"
-  },
+  "repository": "https://github.com/vot/xkpasswd-node.git",
```
Yeah that does make sense. So only corner case is skipping `test` directory from publishing? I agree its not a lot of extra files, so should ok to keep it.

yogendra replied:
Ah didn't see the last line on last comment. I think this is resolved. Let me make these changes.

yogendra committed:
#5 Fixes as per pull comments
Removed .npmignore, .vscode/* and yarn.lock
Updated .gitignore. Added yarn.lock and .vscode for exclusion

yogendra committed:
#5 Narrowed environment (for linting) to nodejs

yogendra committed:
#5 Reverted change to repository in package.json

vot commented:
Thanks @yogendra for making the changes, merging it now.
